content/topics/kotlin/generics/_index.md

```diff
@@ -6,16 +6,342 @@ prerequisites:
+
+The Existential Transformation: Consumer in, Producer out! :-)
+
+### Type projections
+####Use-site variance: Type projections
+It is very convenient to declare a type parameter T as out and avoid trouble with subtyping on the use site, but some classes can't actually be restricted to only return T's! A good example of this is Array:
+```
+class Array<T>(val size: Int) {
+    fun get(index: Int): T { ... }
+    fun set(index: Int, value: T) { ... }
+}
+```
+This class cannot be either co- or contravariant in T. And this imposes certain inflexibilities. Consider the following function:
+```
+fun copy(from: Array<Any>, to: Array<Any>) {
+    assert(from.size == to.size)
+    for (i in from.indices)
+        to[i] = from[i]
+}
+```
+This function is supposed to copy items from one array to another. Let's try to apply it in practice:
+```
+val ints: Array<Int> = arrayOf(1, 2, 3)
+val any = Array<Any>(3) { "" } 
+copy(ints, any)
+//   ^ type is Array<Int> but Array<Any> was expected
+```
+Here we run into the same familiar problem: Array<T> is invariant in T, thus neither of Array<Int> and Array<Any> is a subtype of the other. Why? Again, because copy might be doing bad things, i.e. it might attempt to write, say, a String to from, and if we actually passed an array of Int there, a ClassCastException would have been thrown sometime later.
+
+Then, the only thing we want to ensure is that copy() does not do any bad things. We want to prohibit it from writing to from, and we can:
+```
+fun copy(from: Array<out Any>, to: Array<Any>) { ... }
+```
+What has happened here is called type projection: we said that from is not simply an array, but a restricted (projected) one: we can only call those methods that return the type parameter T, in this case it means that we can only call get(). This is our approach to use-site variance, and corresponds to Java's 
+
+```
+Array<? extends Object>
+```
+but in a slightly simpler way.
+
+
+You can project a type with in as well:
+```
+fun fill(dest: Array<in String>, value: String) { ... }
+```
+Array<in String> corresponds to Java's Array<? super String>, i.e. you can pass an array of CharSequence or an array of Object to the fill() function.
+
+### Star-projections
+Sometimes you want to say that you know nothing about the type argument, but still want to use it in a safe way. The safe way here is to define such a projection of the generic type, that every concrete instantiation of that generic type would be a subtype of that projection.
+
+Kotlin provides so called star-projection syntax for this:
+
+- For 
+
+```
+Foo<out T : TUpper>
+```
+where T is a covariant type parameter with the upper bound TUpper, 
+
+```
+Foo<*> is equivalent to Foo<out TUpper>. 
+```
+
+It means that when the T is unknown you can safely read values of 
+
+```
+TUpper from Foo<*>.
+```
+
+- For 
```

Comment:
some of this should be changed to inline text so it can render as a full paragraph

Reply:
Done